packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+  static const FabPositioner centerFloat = const _CenterFloatFab();
+
+  /// Places the [FloatingActionButton] based on the [Scaffold]'s layout.
+  Offset getOffset(ScaffoldPrelayoutGeometry scaffoldGeometry);
+
+  @override
+  String toString() => '$runtimeType()';
+}
+
+/// Provider of [FloatingActionButton] animations.
+abstract class FabMotionAnimator {
+  const FabMotionAnimator();
+
+  /// Moves the [FloatingActionButton] by scaling out and in at a new location.
+  /// 
+  /// This is the default fab motion animation.
+  static const FabMotionAnimator scaling = const _ScalingFabMotionAnimator();
+
+  /// Animates the [FloatingActionButton]'s [Offset].
+  Animation<Offset> getOffsetAnimation({@required Offset begin, @required Offset end, @required Animation<double> parent});
+
+  /// Animates the scale of the [FloatingActionButton].
+  /// 
+  /// The animation should both start and end with a value of 1.0.
+  Animation<double> getScaleAnimation({@required Animation<double> parent});
+
+  /// Animates the rotation of the [FloatingActionButton].
+  /// 
+  /// The animation should both start and end with a value of 0.0 or 1.0.
+  /// 
+  /// The animation values are a fraction of a full circle, with 0.0 and 1.0
+  /// corresponding to 0 and 360 degrees, while 0.5 corresponds to 180 degrees.
+  Animation<double> getRotationAnimation({@required Animation<double> parent});
+
+  @override
+  String toString() => '$runtimeType';
+}
+
+class _ScalingFabMotionAnimator extends FabMotionAnimator {
+  const _ScalingFabMotionAnimator();
+
+  @override
+  Animation<Offset> getOffsetAnimation({Offset begin, Offset end, Animation<double> parent}) {
+    return new Tween<Offset>(begin: begin, end: end).chain(new CurveTween(curve: const Threshold(0.5))).animate(parent);
+  }
+
+  @override
+  Animation<double> getScaleAnimation({Animation<double> parent}) {
+    final Curve curve = const Interval(0.5, 1.0, curve: Curves.ease);
+    return new _MaxAnimation<double>(
+      new ReverseAnimation(new CurveTween(curve: curve.flipped).animate(parent)),
+      new CurveTween(curve: curve).animate(parent),
+    );
+  }
+
+  @override
+  Animation<double> getRotationAnimation({Animation<double> parent}) {
+    // Because we only see the last half of the rotation tween, 
+    // it needs to go twice as far.
+    final Tween<double> rotationTween = new Tween<double>(
+      begin: 1.0 - _kFloatingActionButtonTurnInterval * 2, 
+      end: 1.0,
+    );
+    // This rotation will turn on the way in, but not on the way out.
+    return new _MaxAnimation<double>(
+      rotationTween.animate(parent),
+      new ReverseAnimation(new CurveTween(curve: const Threshold(0.5)).animate(parent)),
+    );
+  }
+}
+
+class _MaxAnimation<T> extends CompoundAnimation<T> {
+  _MaxAnimation(Animation<T> first, Animation<T> next): super(first: first, next: next);
+
+  @override
+  T get value => math.max(first.value, next.value);
+}
```

Comment:
seems like this would be a potentially useful general animation; should we make it public and put it in the animation library?

Reply:
I think so, but I don't want it to block this PR.  I'll get a separate PR sent to expose MaxAnimation publicly.

Reply:
Sent out in #14814